net/respondd/src/respondd.c

```diff
@@ -41,150 +48,381 @@
 #include <sys/socket.h>
 
 
+struct provider_list {
+	struct provider_list *next;
+
+	char *name;
+	respondd_provider provider;
+};
+
+struct request_type {
+	struct provider_list *providers;
+
+	struct json_object *cache;
+	uint64_t cache_time;
+	int64_t cache_timeout;
+};
+
+
+static int64_t now;
+
+
+static struct json_object * merge_json(struct json_object *a, struct json_object *b);
+
+
 static void usage() {
-  puts("Usage: respondd [-h] -g <group> -p <port> -i <if0> [-i <if1> ..]");
-  puts("  -g <ip6>         multicast group, e.g. ff02::2:1001");
-  puts("  -p <int>         port number to listen on");
-  puts("  -i <string>      interface on which the group is joined");
-  puts("  -c <string>      Lua command");
-  puts("  -h               this help\n");
+	puts("Usage: respondd [-h] -g <group> -p <port> -i <if0> [-i <if1> ..]");
+	puts("	-g <ip6>         multicast group, e.g. ff02::2:1001");
+	puts("	-p <int>         port number to listen on");
+	puts("	-i <string>      interface on which the group is joined");
+	puts("	-d <string>      data provider directory");
+	puts("	-h               this help\n");
 }
 
 static void join_mcast(const int sock, const struct in6_addr addr, const char *iface) {
-  struct ipv6_mreq mreq;
+	struct ipv6_mreq mreq;
 
-  mreq.ipv6mr_multiaddr = addr;
-  mreq.ipv6mr_interface = if_nametoindex(iface);
+	mreq.ipv6mr_multiaddr = addr;
+	mreq.ipv6mr_interface = if_nametoindex(iface);
 
-  if (mreq.ipv6mr_interface == 0)
-    goto error;
+	if (mreq.ipv6mr_interface == 0)
+		goto error;
 
-  if (setsockopt(sock, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq, sizeof(mreq)) == -1)
-    goto error;
+	if (setsockopt(sock, IPPROTO_IPV6, IPV6_JOIN_GROUP, &mreq, sizeof(mreq)) == -1)
+		goto error;
 
-  return;
+	return;
 
  error:
-  fprintf(stderr, "Could not join multicast group on %s: ", iface);
-  perror(NULL);
+	fprintf(stderr, "Could not join multicast group on %s: ", iface);
+	perror(NULL);
+}
+
+
+static void update_time(void) {
+	struct timespec tp;
+	clock_gettime(CLOCK_MONOTONIC, &tp);
+
+	now = (int64_t)tp.tv_sec * 1000 + tp.tv_nsec / 1000000;
+}
+
+
+/** Merges two JSON objects (on conflicts, object a will be preferred) */
+static struct json_object * merge_json(struct json_object *a, struct json_object *b) {
+	if (!json_object_is_type(a, json_type_object) || !json_object_is_type(b, json_type_object)) {
+		json_object_put(b);
+		return a;
+	}
+
+	json_object_object_foreach(a, key, val_a) {
+		struct json_object *val_b;
+
+		json_object_get(val_a);
+
+		if (!json_object_object_get_ex(b, key, &val_b)) {
+			json_object_object_add(b, key, val_a);
+			continue;
+		}
+
+		json_object_get(val_b);
+
+		json_object_object_add(b, key, merge_json(val_a, val_b));
+	}
+
+	json_object_put(a);
+	return b;
+}
+
+static const struct respondd_provider_info * get_providers(const char *filename) {
+	char path[2 + strlen(filename) + 1];
+	snprintf(path, sizeof(path), "./%s", filename);
```

Comment:
Probably everybody apart from me knows that, but I had to look up why `./` is prefixed. Maybe add a comment on that?